text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,187 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It doesn't require javascript since it's only generating links, making it work with or without javascript enabled.
+
+For the link generation, we use the already existing system in rustdoc. Meaning their generation will follow the same rules as the link in the documentation pages.
+
+All items with documentation should have a link to their documentation page (like impl blocks, methods, trait/impl associated items, etc).
+
+So on the technical side, there is no big issues (based on what is described in this RFC). A few parts of this RFC are implemented in the following pull requests:
+
+ * [#84176](https://github.com/rust-lang/rust/pull/84176)
+ * [#88033](https://github.com/rust-lang/rust/pull/88033)
+ * [#91264](https://github.com/rust-lang/rust/pull/91264)
+
+## UI
+
+For the UI, we have the following constraints:
+
+ * Must work on mobile and desktop: we cannot rely on mouse hover events.
+ * It's better if it works without javascript (it's important for accessibility, but also for maintenance reasons too: less JS code, less maintenance).
+
+Some extra explanations about the statement "less JS code, less maintenance": To have this information in the front-end, the back-end has to generate it in any case (either as a link or any other format). So deciding to use JS to handle this would mean that we still have code in the back-end but we would also have code in the front-end to treat it.
+
+With this in mind, the suggested UI would be like this:
+
+ * Links generated on idents in source code will jump to the ident definition.
+ * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
+ * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).
+
+Another thing which would be nice: to make it more obvious if a link points to source code or to documentation, they should have different visual markers (different background/underline colors for example).
+
+Basically, the source code pages' UI doesn't change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**The biggest concern is actually about "scope expansion"**. If we add this feature, are we doing too much? Rustdoc is supposed to be about documentation, so is providing more information into the source code page really necessary?
+
+An answer about this is that it actually makes source code browsing much more pleasant and convenient. Multiple other documentation tools provide this feature for this reason so it doesn't seem like it would be out of scope in this regard.
```

Comment:
This doesn't answer the question. "Rustdoc is about documentation, isn't this scope expansion?" is not answered by "this makes source code browsing more pleasant". In fact, it makes it sound _more_ like scope expansion since you answered a question about one field by mentioning another. It is not at all clear that rustdoc is or should be a tool for browsing the source. Just because it currently _contains_ source is insufficient, it can be a tool that contains source without being a tool for _browsing_ source.

This section (and the motivation section) needs to talk about how source code browsing is related to Rustdoc's job, and why source code browsing is useful for navigating docs.

The thing is, if the docs are good, you shouldn't _need_ to look at the source. This doesn't play out in practice that often but it's an easy argument to make against source code browsing being important. As a community we can (and somewhat have) set our norms towards having docs that don't require that. Would this cause a change to that norm? Would it make people want to write their source code so that people trying to _use_ the library can read it better (this can be a problem! writing a library so that it is readable to maintainers is different from writing one readable to users!).